File: Longest Common Sequence/sequence.cpp
#include <iostream>
#include <cstring>
#include <string>
#include <vector>
#include <algorithm>
#include <random>
#include <chrono>
using namespace std;
using namespace std::chrono;

int mem[1001][1001];

int recursion(string X, string Y, int m, int n)
{
    if (m <= 0 || n <= 0) return 0;

    if (mem[m][n] != -1) return mem[m][n];

    if (X[m-1] == Y[n-1])
    {
        mem[m][n] = 1 + recursion(X, Y, m-1, n-1); 
        return mem[m][n];
    } 

    int left = recursion(X, Y, m, n-1);
    int right = recursion(X, Y, m-1, n);
    mem[m][n] = max(left, right);
    return mem[m][n];
}

int tabular(string X, string Y)
{
    int m = X.size();
    int n = Y.size();
    vector<vector<int>> L(m+1, vector<int>(n+1, 0));

    for (int i=1; i<m+1; i++)
    {
        for (int j=1; j<n+1; j++)
        {
            if (X[i-1] == Y[j-1])
                L[i][j] = 1 + L[i-1][j-1];
            else
                L[i][j] = max(L[i][j-1], L[i-1][j]);
        }
    }
    return L[m][n];
}

int main()
{
    memset(mem, -1, sizeof(mem));
/*
    string X = "ABCBDAB";
    string Y = "BDCABA";
    int m = X.length();
    int n = Y.length();
*/
    int m = 1000;
    int n = 1000;
    string X;
    string Y;
    generate_n (back_inserter(X), m, [] {return static_cast<char> (rand() % 26 + 'A');});
    generate_n (back_inserter(Y), n, [] {return static_cast<char> (rand() % 26 + 'A');});

    auto start = high_resolution_clock::now();    
    //cout << recursion(X, Y, m, n) << endl;
    cout << tabular(X, Y) << endl;
    auto end = high_resolution_clock::now();
    printf("elapsed %.4f seconds.", (end - start) * 1e-9);
}

File: Longest Common Subsequence/sequence.py
def recursion(X, Y, m, n, mem={}):
    
    if m == 0 or n == 0:
        return 0
    
    if (m, n) in mem: return mem[(m, n)]
    
    if X[m-1] == Y[n-1]:
        mem[(m, n)] = 1 + recursion(X, Y, m-1, n-1)
        return mem[(m, n)]
    
    mem[(m, n)] = max(recursion(X, Y, m-1, n), recursion(X, Y, m, n-1))
    return mem[(m, n)]


def tabular(X, Y):
    
    m = len(X)
    n = len(Y)
    L = [[0] * (n+1) for i in range(0, m+1)]
    
    for i in range(1, m+1):
        for j in range(1, n+1):

            if X[i-1] == Y[j-1]:
                L[i][j] = 1 + L[i-1][j-1]
            
            else:
                L[i][j] = max(L[i-1][j], L[i][j-1])

    return L[m][n]


if __name__ == '__main__':

    import string
    import random
    import time
    
    X = 'ABCBDAB'
    Y = 'BDCABA'
    
    A2Z = string.ascii_uppercase
    random.seed(1024)
    X = ''.join([(random.choice(A2Z)) for i in range(1000)])
    Y = ''.join([(random.choice(A2Z)) for i in range(1000)])
    
    start = time.time()
    #print(recursion(X, Y, len(X), len(Y)))
    print(tabular(X, Y))
    end = time.time()
    print('elapsed %.4f seconds' % (end - start))